src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -446,30 +422,25 @@ private void RemoveRootNode()
             // The idea is to replace the specified node by the very last
             // node and shorten the array by one.
 
-            int lastNodeIndex = GetLastNodeIndex();
-            (TElement Element, TPriority Priority) lastNode = _nodes[lastNodeIndex];
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
-            {
-                _nodes[lastNodeIndex] = default;
-            }
+            ref (TElement Element, TPriority Priority) lastNode = ref _nodes[_size - 1];
 
             _size--;
             _version++;
 
             if (_comparer == null)
             {
-                MoveDownDefaultComparer(lastNode, RootIndex);
+                MoveDownDefaultComparer(ref lastNode, 0);
             }
             else
             {
-                MoveDownCustomComparer(lastNode, RootIndex);
+                MoveDownCustomComparer(ref lastNode, 0);
             }
-        }
 
-        /// <summary>
-        /// Gets the index of the last node in the heap.
-        /// </summary>
-        private int GetLastNodeIndex() => _size - 1;
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<(TElement, TPriority)>())
+            {
+                lastNode = default;
+            }
```

Comment:
Why move this to the end?

Reply:
Since it's passed by reference to the MoveDown operation, erasing it beforehand would result in data loss.

Reply:
Ah, I see, ok.